subsys/dfu/img_util/flash_img.c

```diff
@@ -83,3 +85,80 @@ int flash_img_init(struct flash_img_context *ctx)
 {
 	return flash_img_init_id(ctx, FLASH_AREA_IMAGE_SECONDARY);
 }
+
+int flash_img_check_sha256(struct flash_img_context *ctx,
+			   const struct flash_img_check *fic)
+{
+	unsigned char hash[TC_SHA256_DIGEST_SIZE];
+	struct tc_sha256_state_struct sha256sum;
+	int to_read;
+	int pos;
+	int rc;
+
+	if (!fic->match) {
+		return -EINVAL;
+	}
+
+	if (tc_sha256_init(&sha256sum) != TC_CRYPTO_SUCCESS) {
+		return -EIO;
+	}
+
+	to_read = sizeof(ctx->buf);
+
+	for (pos = 0; pos < fic->len; pos += to_read) {
+
+		if (pos + to_read > fic->len) {
+			to_read = fic->len - pos;
+		}
+
+		rc = flash_area_read(ctx->flash_area, fic->off + pos,
+				     ctx->buf, to_read);
+		if (rc != 0) {
+			return rc;
+		}
+
+		if (tc_sha256_update(&sha256sum,
+				     ctx->buf,
+				     to_read) != TC_CRYPTO_SUCCESS) {
+			return -EIO;
+		}
+	}
+
+	if (tc_sha256_final(hash, &sha256sum) != TC_CRYPTO_SUCCESS) {
+		return -EIO;
+	}
+
+	if (memcmp(hash, fic->match, TC_SHA256_DIGEST_SIZE)) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+int flash_img_check(struct flash_img_context *ctx,
+		    const struct flash_img_check *fic,
+		    uint8_t area_id)
+{
+	int rc;
+
+	if (!ctx || !fic) {
+		return -EFAULT;
+	}
+
+	rc = flash_area_open(area_id,
+			     (const struct flash_area **)&(ctx->flash_area));
+	if (rc) {
+		return rc;
+	}
+
+	rc = -EINVAL;
+
+	if (fic->alg == FLASH_IMG_CHECK_ALG_SHA256) {
+		rc = flash_img_check_sha256(ctx, fic);
+	}
```

Comment:
If you would do that check before `flash_are_open`, you could avoid opening in case when algorithm is not set.